### ISSUE TITLE
[macOS] Camera permission status "unknown" after application restart with "granted" status

### ISSUE THREAD

**Issue by itlancer** [airsdk/Adobe-Runtime-Support#889]
## Problem Description
`Camera` permission status "unknown" after application restart with "granted" status with macOS Big Sur using latest AIR.

Tested with latest AIR 33.1.1.476 and multiple macOS Big Sur devices with different Big Sur versions and different acrhitectures (Intel-based and Apple Sillicon M1 based devices) with bundled application.
Application correctly signed and has proper entitlements as described here: https://github.com/airsdk/Adobe-Runtime-Support/issues/208#issuecomment-841883060
With OS system settings camera permission for application checked as "granted".
Same problem in all cases.
After application restart again (when permission was granted) - `Camera.permissionManager.permissionStatus` and `Camera.permissionStatus` both returns "unknown".

Didn't check this with macOS Catalina.
Didn't check the same with Microphone permission.

This bug prevent camera usage for our application cause we cannot request permission after every application restart.

Related issue (not the same): https://github.com/airsdk/Adobe-Runtime-Support/issues/208

## Steps to Reproduce
1) Create bundled AIR application for macOS with code below. Properly sign it and add all necessary entitlements as described here: https://github.com/airsdk/Adobe-Runtime-Support/issues/208#issuecomment-841883060
2) Install and launch application. It trace camera permission status as "unknown" (here it ok). Click anywhere on stage. Application request permission to camera, grant it.
Now Camera permission granted (with OS system setting camera permission for application also will be checked as "granted"). Here all ok.
3) Close application and launch it again. It will trace current camera permission status.

Application example with sources attached.
[macos_camera_permission_bug.zip](https://github.com/airsdk/Adobe-Runtime-Support/files/6591288/macos_camera_permission_bug.zip)

```
package {
	import flash.display.Sprite;
	import flash.events.Event;
	import flash.events.MouseEvent;
	import flash.media.Camera;
	import flash.permissions.PermissionManager;
	import flash.events.PermissionEvent;
	
	public class MacOSCameraPermissionBug extends Sprite {
		
		public function MacOSCameraPermissionBug() {
			addEventListener(Event.ADDED_TO_STAGE, init);
		}
		
		private function init(e:Event):void {
			removeEventListener(Event.ADDED_TO_STAGE, init);
			trace("permissionStatus", Camera.permissionManager.permissionStatus, Camera.permissionStatus);
			
			stage.addEventListener(MouseEvent.CLICK, click);
		}
		
		private function click(e:MouseEvent):void {
			var p:PermissionManager = Camera.permissionManager;
			p.addEventListener(PermissionEvent.PERMISSION_STATUS, permissionStatus);
			p.requestPermission();
		}
		
		private function permissionStatus(e:PermissionEvent):void {
			trace("permissionStatus", e.status);
		}
	}
}
```

**Actual Result:**
`Camera.permissionManager.permissionStatus` and `Camera.permissionStatus` both returns "unknown". Application think that cannot use Camera and need to request permission again.

**Expected Result:**
`Camera.permissionManager.permissionStatus` and `Camera.permissionStatus` both returns "granted".

## Known Workarounds
Write native extension to check real Camera permission status
https://developer.apple.com/documentation/avfoundation/cameras_and_media_capture/requesting_authorization_for_media_capture_on_macos?language=objc
And if you get not "unknown" from native code then real status "granted" or "denied" so you should call `requestPermission` inside AIR to get real status without UI popups.

**Comment by ajwfrost** [comment]
Hi - can I check what actually happens here from a user/UI perspective? From my recollection, the status starts off as "unknown" but when you do a request for the permission, it should then query the OS and immediately respond with "granted" if the user had previously accepted it. So, no new UI or question being raised to the user.

I guess it may be possible for us to bring forward that initial check to within the first query of the status, we can check this..

thanks

**Comment by itlancer** [comment]
@ajwfrost 
On secondary requests seems there is no UI popups and permission granted automatically.
But the point that our application cannot show any popups for secondary requests. It work in public places and showing uncontrollable popups after first set up unacceptable.
`requestPermission` from "unknown" state theoretically cause prompt popup - that we would like to avoid when previously permission already granted (what we cannot detect right now).

It should work as with Android or iOS - getting current permission state without "requesting" that could cause UI popups that cannot be closed or control programmatically.

**Comment by itlancer** [comment]
Found workaround by writing native extension. Approach description added in original post.

**Comment by itlancer** [comment]
Issue still exists with latest AIR 33.1.1.533.

**Comment by ajwfrost** [comment]
Thanks for pointing it out, we'll get a fix into the next release for this..

**Comment by chazhenry** [comment]
Using BigSur 11.1 and Air 33.1.1.554. IntelliJ IDE 'Run' mode. Crashes on pm.requestPermission();  Are the XML and plist.info parameters needed for interactive Run/Debug mode? With the old SDK permission dialogs popped as needed. How do I move forward in dev mode? Thanks.

```
private function cameraSetup():void
	{
	var pm:PermissionManager = Camera.permissionManager;
	trace("Got permission manager for " + pm.resourceType);
	trace("permissionStatus -> " + pm.permissionStatus);
	pm.addEventListener(PermissionEvent.PERMISSION_STATUS, onCameraPermission);
	pm.requestPermission();
	}
```

```
private function onCameraPermission(e : PermissionEvent) : void
	{
	trace("Received PermissionEvent, status = " + e.status);
	var names : Array = Camera.names;
	trace("Found " + names.length + " cameras");
	for (var i : uint = 0; i < names.length; i++)
		{
		trace(i + ": " + names[i]);
		}
	if (e.status == PermissionStatus.GRANTED) startCamera();
	}
```

**Comment by ajwfrost** [comment]
@chazhenry what you're seeing there is the same issue as #995 and #208  - basically, ADL is being launched as a child process by Animate/IDEA/etc and the permissions for things like camera/mic access are then taken from the parent process. We've just been investigating this one, and I don't think it's going to be possible for us to fix it: whatever we try to do, the way macOS and its security works means that it's Animate/IDEA/etc that need to have the correct entitlements and information strings...

... bit of an update though - we have a hack for this, to set up the OS to allow it...

First, open the permissions database:
```
$ sqlite3 ~/Library/Application\ Support/com.apple.TCC/TCC.db
```
And then, add this manually:
```
insert into access (service, client, client_type, auth_value, auth_reason, auth_version, flags) values ('kTCCServiceMicrophone', 'com.adobe.Adobe-Animate-2021.application', 0, 2, 2, 1, 0);
.quit
```
and check that it then appears in the permissions...

You can do the same with other bundle IDs for IntelliJ or whatever (find the bundle ID from their Info.plist file within the application bundle contents...), and for camera it's `kTCCServiceCamera`

Perhaps not a very official way of doing it, but it seems to now allow it to execute properly. Potentially we can/should do this programmatically within ADL...!

**Comment by chazhenry** [comment]
@ajwfrost I appreciate the quick response. No joy trying the sqlite3 edit (authorization denied). 

I've plowed through and compiled a captive runtime bundle. When I run that app, It first traces the 'unknown' permission status as before, but when it calls pm.requestPermission(); I get an immediate 'denied' inside the permission event handler - no dialog or chance to approve.

**Comment by ajwfrost** [comment]
@chazhenry curious - are you able to look at the permissions via "System Preferences" -> "Security & Privacy" -> "Privacy" tab and "Camera" in the list. If you're getting an immediate "denied" then it should be showing up in here...

On the sqlite3 thing, is the error in the initial "sqlite3" command, or when you try to run the 'insert' operation? Might be worth adding "sudo" at the start of the sqlite3 command perhaps; but regardless, if you can get into the database then try running a simple query e.g. `select * from access where service = 'kTCCServiceCamera';` and see if it lists anything..

**Comment by chazhenry** [comment]
@ajwfrost I checked SystemPrefs>Security>Privacy>Camera as I was running - only GoogleChrome is authorized for Camera. I don't see a way to add an app until it shows up there requesting which my app doesn't seem to do. 

For sqlite3, I get the 'auth denied' on the sqlite command itself. Was running as root via sudo -s. 

I'm a Windows dev and just poking with a stick at the Mac. Isn't my app compiled using the XML and plist.info? Could it be in there that I need to request permissions? Mac seems incredibly redundant and restrictive. Seems that app should be able to request on the fly. Thanks again.

**Comment by chazhenry** [comment]
@ajwfrost  Looking at #532 - am I understanding that I'll need to Build a captive runtime bundle and then codesign it to test the camera and mic element of my app? Codesign has never worked in the past, but I also admit that I don't understand the relation between the app.xml file and info.plist. I was under the impression that the build process created info.plist but I don't know the syntax of app.xml that would make the necessary entitlements flow into it. Everything is working in Windows - lost in Mac hell.

**Comment by ajwfrost** [comment]
Hi @chazhenry  - yes it's a bit of a pain... but to get it to work, you'll need both the Info.plist updates - which you will get if you've added that `InfoAdditions` line per your earlier post - and then to codesign it with the appropriate entitlements. I'm pretty sure we also support entitlements to be added from the XML descriptor so it should be that you can add:
```
<macOS>
<InfoAdditions>
    <![CDATA[
        <key>NSMicrophoneUsageDescription</key>
        <string>Bob needs to access the mic for some reason</string>
    ]]>
</InfoAdditions>
<Entitlements>
    <![CDATA[
        <key>com.apple.security.device.camera</key>
        <true/>
    ]]>
</Entitlements>
</macOS>
```

And then to do the code-signing with those entitlements, run ADL using:
```
adt -target [signing options] -package bundle -storetype keystore -alias "Developer ID Application" MyApplication.app application.xml etc etc
```
i.e. as well as the normal AIR signing options, you can provide a certificate name that will then be used with the `codesign` tool (as the "-s" parameter) and those entitlements should then be injected.

You can check the final signed app using:
```
codesign -d --entitlements - MyApplication.app
```

Hope that helps...!

**Comment by chazhenry** [comment]
@ajwfrost Still don't see where/if my app creates an info.plist file.

**Comment by chazhenry** [comment]
@ajwfrost First, thanks for the detailed info. Wanted to get that out before saying that I'm likely dropping out. I've been evaluating Harman as way to keep an Air app alive. The progress seems steady, but slow. When the latest release fixed my Windows camera issues, I was ready to subscribe. 

Then these Mac issues. Honestly, coming from a dev env that can hotload code changes, this is like going backward 20 years. No serious dev environment requires you to compile and codesign to run/test. I have a camera intensive app and a hundred tweaks yet to make. Can't imagine doing a full release cycle to test each change. I no longer see this as a viable cross platform dev env. Disillusioned, but appreciative for the help which made clear the state of the tools.

**Comment by ajwfrost** [comment]
@chazhenry Part of our problem is that there are third party tools (IntelliJ/Animate/etc) that are kicking off the ADL process from the AIR SDK - and from Apple's security model, they then look at the permissions given to the original app (IntelliJ in your case) rather than ADL. So - debugging anything with a camera in it looks to be impossible if you're using IntelliJ/Animate/etc. We're looking at a workaround where you'd need to initially load one of our processes and then we'd change ADL so that it just kicks this process to load the files, which should work around this macOS restriction... or alternatively, we'd have to set everything up to use Xcode which of course does all the building/code-signing nice and quietly behind the scenes...

Sorry, there was a typo in my earlier comment where I say to run ADL, I meant ADT (i.e. the command line was right, my intro line was wrong). ADT = the AIR Developer Tool which is what you use (or IntelliJ uses) to build the AIR application in the first place. So, if you package up the application using ADT, it gives you a bit more control over things rather than having to just do whatever these third party tools are set up to do.

Once you have created your application, it will be a ".app" bundle - which is just a folder. If you "Show package contents" in Finder - or just change into that folder if you're using a terminal - then you will see a Contents folder within which is the Info.plist file. So at that point, you should see the permission strings within Info.plist. ADL (AIR Debug Launcher) has the Info.plist built into it, although when I run ADL from a terminal window, it's the Terminal process that ends up in the list of apps that have been granted camera positions. Probably again due to how macOS considers the security... but if you double-click on the .app file to launch it from Finder, then it would be this process that's actually requesting the permission.

Basically - yes, really convoluted, and not helped by the fact that Apple update their requirements and everything then lags behind i.e. we put changes into the AIR SDK but it will take time for IntelliJ/Animate/etc to update their applications to handle this hence these workarounds...

**Comment by chazhenry** [comment]
@ajwfrost Thanks for the detail. I can confirm that none of the InfoAdditions or Entitlements that I put in the app.xml file are making their way to the Info.plist file (and thus the compiled app). Assuming that I can't ever run in debug mode, it would be a huge half step to compile and run instead of compile, codesign, and run. Still, sounds like I can't get there from here. Is it an IntelliJ issue? If so, is there a script that runs ADL - what should I be using instead? Command line? Thanks.

**Comment by ajwfrost** [comment]
@chazhenry one thing that may be required in order to get those attributes into the Info.plist file is to have the namespace set at 33.1 - and to be using a recent release of AIR SDK: 33.1.1.406 or later, by the looks of things. Although .. I would expect, if either of those weren't the case, then the toolchain would just complain about unknown attributes etc.

Ultimately there are three steps to all this:

1) Compile. This is sometimes a bit tricky to get all the command-line arguments set up properly for this so it's useful to have IntelliJ for this.. I don't know exactly how IntelliJ works but for a command-line build, you'd use the "mxmlc" script that's in the SDK's 'bin' folder: so you could perhaps edit that script and output all command-line arguments to a file, so that you can see what's going on and then replicate it. Although, it may be that IntelliJ bypasses that script and just calls the Java application directly....

2) Run in the debugger. So for this IntelliJ (or whatever) would call ADL, and this is where the problem lies in terms of permissions. To get round this you can just call ADL directly, and it's a bit simpler here. So if you leave IntelliJ to do the compilation, you should end up with an application descriptor XML file, and a main SWF file. Then you can just call from a command line/terminal window:
`adl application.xml path_to_swf`
so if your XML and SWF are in the same place and you're in the same folder as them:
`/sdk_path/bin/adl application.xml .`
(you may even be able to omit the '.' at the end .. just means 'look for the SWF in the current folder').
This should allow you to run the app. To actually debug may be a bit more complex but if there's any "start remote debug session" option in IntelliJ, then try that: it should put IntelliJ into a "waiting for debugger" state, without it actually trying to launch the ADL process itself. So then when you do it, it would connect to the waiting debugger (you have about 30s I think).

3) Package. So this is where you are using ADT, and creating a bundle for your application, which would involve the creation of the Info.plist file. This should create the .app folder with Contents/Info.plist and Contents/Frameworks/Adobe\ AIR.framework and should then be what you distribute....

Hope that helps; we'll see what we can do to make some of this a bit easier!

**Comment by chazhenry** [comment]
@ajwfrost Namespace in the XML file is below. Full file attached (renamed to txt for attachment). 
`<application xmlns="http://ns.adobe.com/air/application/33.1">`

Using 33.1.1.554. Can compile and package. Production App runs and displays the Harman splash. No sign of <infoadditions> or <entitlements> in the plist file from Contents and no permission dialogs popped at runtime.
[Info.plist.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/6887224/Info.plist.txt)
[app.xml.txt](https://github.com/airsdk/Adobe-Runtime-Support/files/6887225/app.xml.txt)

Will try command line, but would love to get the IDE tuned as well.

**Comment by itlancer** [comment]
Original issues with camera permission "unknown" status after application restart fixed with latest AIR 33.1.1.575.
Thanks!
